### PR TITLE
[Az] Fix get token code to consume latest Az.Accounts module (2.2.8)

### PR DESCRIPTION
**Pull request by rakku-ms**

### CODE REVIEW

File: Registration/RegisterWithAzure.psm1 [hunk 1]
```diff
@@ -1651,54 +1647,41 @@ function Get-AzToken
...
+            $accounts = $client.GetAccountsAsync().ConfigureAwait($true).GetAwaiter().GetResult()
+            $bytes = ([Microsoft.Identity.Client.ITokenCacheSerializer]$client.UserTokenCache).SerializeMsalV3()
+        }
+        else
+        {
+            $provider = [Microsoft.Azure.Commands.Common.Authentication.SharedTokenCacheProvider]::new()
+            $accounts = $provider.ListAccounts()
+            $bytes = $provider.ReadTokenData()
+        }
         $json  = [System.Text.Encoding]::UTF8.GetString($bytes)
```
Comment (keystroke) [inline Registration/RegisterWithAzure.psm1:1685]: let's maybe add a check here for $bytes.Length and give some error message if its empty?

File: Registration/RegisterWithAzure.psm1 [hunk 2]
```diff
@@ -1610,9 +1610,9 @@ function DeActivate-AzureStack{
     } while ($currentAttempt -lt $maxAttempt)
 }
 
-function Get-AzToken
+function Export-AzRefreshToken
 {
-    [CmdletBinding(DefaultParameterSetName='default')]
+    [CmdletBinding()]
     param
```
Comment (keystroke) [inline Registration/RegisterWithAzure.psm1:1616]: can we remove "resource" param as well?

Comment (rakku-ms) [reply Registration/RegisterWithAzure.psm1:1616]: updated

File: Registration/RegisterWithAzure.psm1 (hunk 1 above)
Comment (rakku-ms) [reply Registration/RegisterWithAzure.psm1:1684]: updated

File: Registration/RegisterWithAzure.psm1 [hunk 3]
```diff
@@ -1649,56 +1640,47 @@ function Get-AzToken
...
-            [Microsoft.Azure.Commands.Common.Authentication.Authentication.Clients.AuthenticationClientFactory]::AuthenticationClientFactoryKey,
-            [ref]$authenticationClientFactory)))
+        if (-not $bytes)
         {
-            $m = 'Please ensure you have authenticated with Az Accounts module!'
-            $m += ' Unable to resolve authentication client factory from Az Accounts module runtime'
-            $m += ' ([Microsoft.Azure.Commands.Common.Authentication.Authentication.Clients.AuthenticationClientFactory])'
-            Write-Error $m
-            return
+            Write-Error "Ensure you enable context autosave for the process (using Enable-AzContextAutosave -Scope Process) and then login before calling this method."
```
Comment (keystroke) [inline Registration/RegisterWithAzure.psm1:1682]: need to return after this

Comment (keystroke) [reply Registration/RegisterWithAzure.psm1:1682]: Error: unable to resolve refresh token from empty context. Ensure you enable.....<>

Comment (rakku-ms) [reply Registration/RegisterWithAzure.psm1]: fixed

**Review by keystroke — APPROVED** [review]

File: Registration/RegisterWithAzure.psm1 [hunk 4]
```diff
@@ -1649,56 +1640,48 @@ function Get-AzToken
...
+                ($useAdfs=$Context.Environment.ActiveDirectoryAuthority -like '*/adfs*'))
+
+            $authenticationClientFactory.RegisterCache($client)
+            $accounts = $client.GetAccountsAsync().ConfigureAwait($true).GetAwaiter().GetResult()
+            $bytes = ([Microsoft.Identity.Client.ITokenCacheSerializer]$client.UserTokenCache).SerializeMsalV3()
         }
         else
         {
-            Write-Warning "Attempting to find a refresh token and an access token from the existing token cache data..."
+            $provider = [Microsoft.Azure.Commands.Common.Authentication.SharedTokenCacheProvider]::new()
```
Comment (bganapa) [inline Registration/RegisterWithAzure.psm1:1676]: Do we need a similar change in identity scripts?

Comment (rakku-ms) [reply Registration/RegisterWithAzure.psm1:1676]: yes - https://github.com/Azure/AzureStack-Tools/blob/46071b431910a4157ce9ac35697876bbf3676c47/Identity/AzureStack.Identity.Common.psm1#L111

we can't remove the accesstoken part in the identity script as it uses access token too

Comment (keystroke) [reply Registration/RegisterWithAzure.psm1:1676]: I think the preference would be to use the new Get-AzsToken command (from the new az module) separately for access tokens, and only use the export-refreshtoken function for the refresh token. In the identity script, perhaps we can keep the "aggregate" function on the top-level which can turn-around and call the other two functions ionternally, along with whatever fallback for backwards compatibility that is required?

Comment (rakku-ms) [reply Registration/RegisterWithAzure.psm1:1676]: Your new suggested approach can get the access token from cache too. I just updated the function in the identity common script. can you please take a look at this commit https://github.com/Azure/AzureStack-Tools/pull/600/commits/1696b224d484dd6c0451e8ba7a7f1a7ff9be2620

**Review by bganapa — APPROVED** [review]

**Pull request merged**